batch/batch/driver/instance_collection/pool.py

```diff
@@ -323,28 +323,36 @@ async def regions_to_ready_cores_mcpu_from_estimated_job_queue(self) -> List[Tup
   SELECT scheduling_iteration, user_idx, n_regions, regions_bits_rep, CAST(COALESCE(SUM(cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
   FROM (
     SELECT {user_idx} AS user_idx, batch_id, job_id, cores_mcpu, always_run, n_regions, regions_bits_rep,
-      ROW_NUMBER() OVER (ORDER BY batch_id, always_run DESC, -n_regions DESC, regions_bits_rep, job_id ASC) DIV {share} AS scheduling_iteration
+      ROW_NUMBER() OVER (ORDER BY batch_id, job_group_id, always_run DESC, -n_regions DESC, regions_bits_rep, job_id ASC) DIV {share} AS scheduling_iteration
     FROM (
       (
-        SELECT jobs.batch_id, jobs.job_id, cores_mcpu, always_run, n_regions, regions_bits_rep
+        SELECT jobs.batch_id, jobs.job_id, jobs.job_group_id, cores_mcpu, always_run, n_regions, regions_bits_rep
         FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
         LEFT JOIN batches ON jobs.batch_id = batches.id
         WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND always_run AND inst_coll = %s
-        ORDER BY jobs.batch_id ASC, jobs.job_id ASC
+        ORDER BY jobs.batch_id ASC, jobs.job_group_id ASC, jobs.job_id ASC
         LIMIT {share * self.job_queue_scheduling_window_secs}
       )
       UNION
       (
-        SELECT jobs.batch_id, jobs.job_id, cores_mcpu, always_run, n_regions, regions_bits_rep
+        SELECT jobs.batch_id, jobs.job_id, jobs.job_group_id, cores_mcpu, always_run, n_regions, regions_bits_rep
         FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
         LEFT JOIN batches ON jobs.batch_id = batches.id
-        LEFT JOIN job_groups_cancelled ON batches.id = job_groups_cancelled.id
-        WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND NOT always_run AND job_groups_cancelled.id IS NULL AND inst_coll = %s
-        ORDER BY jobs.batch_id ASC, jobs.job_id ASC
+        LEFT JOIN LATERAL (
+          SELECT 1 AS cancelled
+          FROM job_group_self_and_ancestors
+          INNER JOIN job_groups_cancelled
+          ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
+             job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
+          WHERE jobs.batch_id = job_group_self_and_ancestors.batch_id AND
+                jobs.job_group_id = job_group_self_and_ancestors.job_group_id
+        ) AS t ON TRUE
```

Comment:
Same comment about upward traversals versus just setting it once.

Reply:
See https://github.com/hail-is/hail/pull/14282#discussion_r1487618770